File: src/regulae/diagnostics.py
"""Diagnostic helpers over a trained :class:`MultiLectModel`.

These functions consume an already-trained model and the corpus it
was trained on, and emit human-oriented reports rather than
modifying the model. Nothing here feeds back into training,
scoring, or alignment — the framework's behavior is unchanged
whether or not any of these are called.

Current inventory:

* :func:`find_cognate_outliers` — rank cognate sets by
  alignment-cost anomaly so the user can spot likely non-cognates,
  loanwords, transcription errors, or genuinely atypical regular
  forms.

-- On cognate input --

The framework uses an explicit-cognate input model: it does **not** try to
infer cognacy from glosses or wordlists, and does **not** filter
out pairs that look non-cognate during training. Every pair the
user passes in is trained on. This is a deliberate choice: glosses
are unreliable cognacy markers, and automatic cognate-detection
belongs to a different tool, not to the alignment engine.

The consequence is that if you pass the framework a pair list
that contains non-cognates — e.g., the classic ``kaput`` /
``kabeθa`` (Latin *caput* "head" vs Spanish "head" < *capitia*)
— those pairs get trained on and contribute observations to the
segment table. The framework has no way to know they shouldn't
be there.

:func:`find_cognate_outliers` is a **post-hoc diagnostic**: after
training, it ranks the input cognate sets by how well their forms
actually align under the model. Pairs whose alignment cost is far
above the corpus mean are flagged as outliers. The user then
decides what to do about them: drop them and retrain, keep them
with a note, or accept them as genuine regular-but-rare
correspondences.

This is a tool, not a gate. The framework stays honest: we never
silently reject input.
"""

import itertools
import math
from collections.abc import Sequence
from dataclasses import dataclass

from regulae.model import CognateSet, LearnedModel, MultiLectModel
from regulae.search import align_forms, alignment_cost


@dataclass(frozen=True)
class CognateOutlierReport:
    """One row of a cognate outlier ranking.

    * ``cognate_id`` — identifier of the input cognate set.
    * ``n_pairs`` — number of lect pairs whose alignment was
      computed for this set (``N*(N-1)/2`` for an N-lect set,
      minus any pairs with no trained model).
    * ``cost_per_segment`` — mean alignment cost per segment
      averaged over all contributing pair alignments for this
      cognate set. Length-normalized so short words aren't
      mechanically cheaper than long ones.
    * ``z_score`` — how many standard deviations above the
      corpus mean the ``cost_per_segment`` sits. Positive
      values are the outliers; strong outliers have z > 2.
    """

    cognate_id: str
    n_pairs: int
    cost_per_segment: float
    z_score: float


def _pair_cost_per_segment(
    form_a, form_b, model: LearnedModel, max_chunk_size: int
) -> float:
    """Compute length-normalized alignment cost for one pair.

    Cost normalization divides the total alignment cost by
    ``(len(form_a) + len(form_b)) / 2``. Short cognates and long
    cognates are compared on the same scale.
    """
    alignment = align_forms(
        form_a, form_b, max_chunk_size=max_chunk_size, model=model
    )
    cost = alignment_cost(alignment, model=model)
    denom = (len(form_a.segments) + len(form_b.segments)) / 2.0
    if denom <= 0.0:
        return 0.0
    return cost / denom


def find_cognate_outliers(
    corpus: Sequence[CognateSet],
    model: MultiLectModel,
    *,
    top_k: int | None = None,
    max_chunk_size: int = 3,
) -> list[CognateOutlierReport]:
    """Rank cognate sets in the corpus by alignment-cost anomaly.

    For each cognate set, every pair of lects present in both the
    set and the model's ``pairwise_models`` is aligned, and the
    length-normalized per-segment alignment cost is averaged
    across those pairs. Cognate sets are then sorted by a z-score
    against the corpus-wide mean and standard deviation of that
    statistic.

    Arguments:

    * ``corpus``: the cognate sets the model was trained on (or
      any cognate sets you'd like to score under this model).
    * ``model``: a :class:`MultiLectModel` whose
      ``pairwise_models`` covers the lects in the corpus.
    * ``top_k``: if given, return only the top ``k`` entries by
      z-score (the most suspect sets). If ``None`` (default),
      return every set that had at least one computable pair.
    * ``max_chunk_size``: forwarded to :func:`align_forms`.

    Returns a list of :class:`CognateOutlierReport`, sorted by
    z-score descending (most anomalous first).

    Interpretation:

    * ``z_score`` near 0 means the set aligns about as cheaply as
      a typical set in this corpus — nothing unusual.
    * ``z_score > 2`` is traditionally "significantly above the
      mean". These are candidates for inspection: likely
      non-cognates, loanwords, transcription errors, or
      genuinely unusual regular sets.
    * Negative ``z_score`` means the set aligns cheaper than
      average. These are the boring, confirmatory sets — the
      bread and butter of the correspondence evidence.

    The function is a **diagnostic**, not a filter. It does not
    modify the model or the corpus. It's meant to be called
    once after training to decide which sets deserve a second
    look from a human or a separate reviewer.
    """
    scored: list[tuple[str, int, float]] = []  # (id, n_pairs, cost_per_seg)
    for cs in corpus:
        lects = sorted(cs.forms)
        total = 0.0
        n_pairs = 0
        for lect_a, lect_b in itertools.combinations(lects, 2):
            key = frozenset({lect_a, lect_b})
            pair_model = model.pairwise_models.get(key)
            if pair_model is None:
                continue
            c = _pair_cost_per_segment(
                cs.forms[lect_a],
                cs.forms[lect_b],
                pair_model,
                max_chunk_size,
            )
            total += c
            n_pairs += 1
        if n_pairs == 0:
            continue
        avg = total / n_pairs
        scored.append((cs.cognate_id, n_pairs, avg))

    if not scored:
        return []

    # Compute mean and standard deviation of the per-segment costs.
    # For small corpora (n < 2) std is undefined; treat it as 0 and
    # return z-scores of 0 (nothing is an outlier if we have no
    # distribution to compare against).
    costs = [c for _, _, c in scored]
    n = len(costs)
    mean = sum(costs) / n
    if n < 2:
        std = 0.0
    else:
        variance = sum((c - mean) ** 2 for c in costs) / (n - 1)
        std = math.sqrt(variance)

    reports: list[CognateOutlierReport] = []
    for cid, npairs, cost in scored:
        z = (cost - mean) / std if std > 0.0 else 0.0
        reports.append(
            CognateOutlierReport(
                cognate_id=cid,
                n_pairs=npairs,
                cost_per_segment=cost,
                z_score=z,
            )
        )

    reports.sort(key=lambda r: (-r.z_score, r.cognate_id))
    if top_k is not None:
        reports = reports[:top_k]
    return reports

File: diagnostics.go
package regulae

import (
	"math"
	"sort"
)

// Diagnostic helpers over a trained MultiLectModel. These consume a trained
// model and emit human-oriented reports; nothing here feeds back into training,
// scoring, or alignment.

// CognateOutlierReport is one row of a cognate outlier ranking. ZScore is how
// many standard deviations above the corpus mean the length-normalized
// alignment cost sits; positive values (especially > 2) are the outliers.
type CognateOutlierReport struct {
	CognateID      string
	NPairs         int
	CostPerSegment float64
	ZScore         float64
}

// pairCostPerSegment computes the length-normalized alignment cost for one
// pair (total cost divided by the mean form length).
func pairCostPerSegment(formA, formB Form, model *LearnedModel, maxChunkSize int) float64 {
	alignment := alignForms(formA, formB, model.FeatureSystem, maxChunkSize, model)
	cost := alignmentCost(alignment, model.FeatureSystem, model)
	denom := float64(len(formA.Segments)+len(formB.Segments)) / 2.0
	if denom <= 0.0 {
		return 0.0
	}
	return cost / denom
}

// FindCognateOutliers ranks cognate sets by alignment-cost anomaly (z-score of
// the per-segment cost against the corpus mean), most anomalous first. topK <= 0
// returns every set with at least one computable pair. It is a diagnostic, not
// a filter. Returns an *UnknownGraphemeError on an unrecognised grapheme.
func FindCognateOutliers(corpus []CognateSet, model *MultiLectModel, topK, maxChunkSize int) (reports []CognateOutlierReport, err error) {
	defer catchUnknownGrapheme(&err)
	if maxChunkSize == 0 {
		maxChunkSize = DefaultMaxChunkSize
	}

	type scoredEntry struct {
		id     string
		nPairs int
		cost   float64
	}
	var scored []scoredEntry
	for _, cs := range corpus {
		lects := make([]string, 0, len(cs.Forms))
		for l := range cs.Forms {
			lects = append(lects, l)
		}
		sort.Strings(lects)
		total := 0.0
		nPairs := 0
		for i := 0; i < len(lects); i++ {
			for j := i + 1; j < len(lects); j++ {
				pm, ok := model.PairwiseModel(lects[i], lects[j])
				if !ok {
					continue
				}
				total += pairCostPerSegment(cs.Forms[lects[i]], cs.Forms[lects[j]], pm, maxChunkSize)
				nPairs++
			}
		}
		if nPairs == 0 {
			continue
		}
		scored = append(scored, scoredEntry{id: cs.CognateID, nPairs: nPairs, cost: total / float64(nPairs)})
	}

	if len(scored) == 0 {
		return nil, nil
	}

	n := len(scored)
	mean := 0.0
	for _, s := range scored {
		mean += s.cost
	}
	mean /= float64(n)
	std := 0.0
	if n >= 2 {
		variance := 0.0
		for _, s := range scored {
			d := s.cost - mean
			variance += d * d
		}
		variance /= float64(n - 1)
		std = math.Sqrt(variance)
	}

	reports = make([]CognateOutlierReport, 0, len(scored))
	for _, s := range scored {
		z := 0.0
		if std > 0.0 {
			z = (s.cost - mean) / std
		}
		reports = append(reports, CognateOutlierReport{CognateID: s.id, NPairs: s.nPairs, CostPerSegment: s.cost, ZScore: z})
	}
	sort.SliceStable(reports, func(i, j int) bool {
		if reports[i].ZScore != reports[j].ZScore {
			return reports[i].ZScore > reports[j].ZScore
		}
		return reports[i].CognateID < reports[j].CognateID
	})
	if topK > 0 && len(reports) > topK {
		reports = reports[:topK]
	}
	return reports, nil
}
